sklearn/metrics/_dist_metrics.pyx

```diff
@@ -554,27 +546,49 @@ cdef class MinkowskiDistance(DistanceMetric):
     r"""Minkowski Distance
 
     .. math::
-       D(x, y) = [\sum_i |x_i - y_i|^p] ^ (1/p)
+       D(x, y) = [\sum_i w_i *|x_i - y_i|^p] ^ (1/p)
+
+    Parameters
+    ----------
+    p : int
+        The order of the norm of the difference :math:`{||u-v||}_p`.
+    w : (N,) array-like (optional)
+        The weight vector.
 
     Minkowski Distance requires p >= 1 and finite. For p = infinity,
     use ChebyshevDistance.
     Note that for p=1, ManhattanDistance is more efficient, and for
     p=2, EuclideanDistance is more efficient.
     """
-    def __init__(self, p):
+    def __init__(self, p, w=None):
         if p < 1:
             raise ValueError("p must be greater than 1")
         elif np.isinf(p):
             raise ValueError("MinkowskiDistance requires finite p. "
                              "For p=inf, use ChebyshevDistance.")
+        elif w is not None and any(w_i < 0 for w_i in w):
+            raise ValueError("w cannot contain negative weights")
+
         self.p = p
+        self.vec = ReadonlyArrayWrapper(np.asarray([], dtype=DTYPE))
+        self.size = 0
```

Comment:
The code would be more intuitive (and slightly more efficient) with those two lines in the `else` clause of the condition below.